### PR TITLE
Add app services registration (dependency injection)

### PR DESCRIPTION
**Pull request by Hatles**
## Proposed change

- [ ] Register services for dynamically loaded assemblies

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code compiles without warnings (code quality chek)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/net-daemon/docs

### CODE REVIEW

File: src/App/NetDaemon.App/Common/NetDaemonAppServicesAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+
+namespace NetDaemon.Common
+{
+    /// <summary>
+    /// Marks a class as a NetDaemonApp
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class)]
```
Comment (FrankBakkerNl) [inline src/App/NetDaemon.App/Common/NetDaemonAppServicesAttribute.cs:8]: Not sure what this attribute is for the comment seem to match

File: src/Daemon/NetDaemon.Daemon/Daemon/CodeServicesManager.cs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+        
+        public IServiceCollection ConfigureServices(IServiceCollection services)
+        {
+            var serviceProvider = services.BuildServiceProvider();
+            // Instantiate all service configuration classes
+            var servicesInstances = InstantiateServicesTypes(serviceProvider);
+
+            // Add each app as a service
+            foreach (var appType in DaemonAppTypes)
```
Comment (FrankBakkerNl) [inline src/Daemon/NetDaemon.Daemon/Daemon/CodeServicesManager.cs:51]: Why do we actually need apps a services? Some apps can be multy instance so a singleton would not work in those cases

File: src/Daemon/NetDaemon.Daemon/Daemon/CodeServicesManager.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                    }
+
+                    throw new Exception("app not found");
+                });
+            }
+            
+            // Add daemon app services
+            foreach (object appServices in servicesInstances)
+            {
+                var configFn = appServices.GetType().GetMethod("ConfigureServices", BindingFlags.Public | BindingFlags.Instance);
```
Comment (FrankBakkerNl) [inline src/Daemon/NetDaemon.Daemon/Daemon/CodeServicesManager.cs:72]: ConfigureServices could be on any class not just apps.

File: src/Daemon/NetDaemon.Daemon/Daemon/CodeServicesManager.cs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+            }
+            
+            public object? GetService(Type serviceType)
+            {
+                if (serviceType == typeof(IServiceScopeFactory))
+                {
+                    return this;
+                }
+                
+                var appsService = _appsProvider.GetService(serviceType);
```
Comment (FrankBakkerNl) [inline src/Daemon/NetDaemon.Daemon/Daemon/CodeServicesManager.cs:109]: Does this works when an app needs a service from the 'parent' provider?

File: src/App/NetDaemon.App/Common/NetDaemonAppServicesAttribute.cs (hunk 1 above)
Comment (BeeHiveJava) [reply src/App/NetDaemon.App/Common/NetDaemonAppServicesAttribute.cs:8]: This makes it so this attribute (`NetDaemonAppServicesAttribute`) is only valid on `class` declarations.

Comment (FrankBakkerNl) [reply src/App/NetDaemon.App/Common/NetDaemonAppServicesAttribute.cs:8]: I ment what this atrribute itself is supposed to do

File: src/Daemon/NetDaemon.Daemon/Daemon/CodeServicesManager.cs (hunk 3 above)
Comment (Hatles) [reply src/Daemon/NetDaemon.Daemon/Daemon/CodeServicesManager.cs]: servicesInstances are classes registered as services providers (aka with the new attribute describe below: NetDaemonServicesProviderAttribute)

File: src/App/NetDaemon.App/Common/NetDaemonAppServicesAttribute.cs (hunk 1 above)
Comment (Hatles) [reply src/App/NetDaemon.App/Common/NetDaemonAppServicesAttribute.cs:8]: I updated doc/comment and renamed the attribute to NetDaemonServicesProviderAttribute

File: src/Daemon/NetDaemon.Daemon/Daemon/CodeServicesManager.cs (hunk 4 above)
Comment (Hatles) [reply src/Daemon/NetDaemon.Daemon/Daemon/CodeServicesManager.cs]: I removed this code, as this was an experiment about child and parent service providers. But it doesn't work well with scopes.

File: src/Daemon/NetDaemon.Daemon/Daemon/CodeServicesManager.cs (hunk 2 above)
Comment (Hatles) [reply src/Daemon/NetDaemon.Daemon/Daemon/CodeServicesManager.cs]: As now, we can register services, apps can communicate through those services. So you are right, we can remove registering of apps as services.

**Comment by helto4real** [comment]
@FrankBakkerNl is this PR obsolete now regarding the previous one?

**Pull request closed without merging**